File: typescript/@relateby/pattern/src/standard-graph.ts
import { Effect, Option, pipe } from "effect"
import { Gram } from "./gram.js"
import { GramParseError } from "./errors.js"
import { Pattern } from "./pattern.js"
import { Subject } from "./subject.js"

export interface StandardRelationship {
  readonly pattern: Pattern<Subject>
  readonly source: string
  readonly target: string
}

export class StandardGraph {
  private readonly _nodes = new Map<string, Pattern<Subject>>()
  private readonly _relationships = new Map<string, StandardRelationship>()
  private readonly _annotations = new Map<string, Pattern<Subject>>()
  private readonly _walks = new Map<string, Pattern<Subject>>()
  private readonly _other = new Map<string, Pattern<Subject>>()

  static fromPatterns(patterns: ReadonlyArray<Pattern<Subject>>): StandardGraph {
    const graph = new StandardGraph()
    for (const pattern of patterns) {
      graph.ingest(pattern)
    }
    return graph
  }

  static fromGram(input: string): Effect.Effect<StandardGraph, GramParseError> {
    return pipe(Gram.parse(input), Effect.map(StandardGraph.fromPatterns))
  }

  get nodeCount(): number {
    return this._nodes.size
  }

  get relationshipCount(): number {
    return this._relationships.size
  }

  get annotationCount(): number {
    return this._annotations.size
  }

  get walkCount(): number {
    return this._walks.size
  }

  get isEmpty(): boolean {
    return this.nodeCount === 0 &&
      this.relationshipCount === 0 &&
      this.annotationCount === 0 &&
      this.walkCount === 0 &&
      this._other.size === 0
  }

  get hasConflicts(): boolean {
    return false
  }

  nodes(): IterableIterator<[string, Pattern<Subject>]> {
    return this._nodes.entries()
  }

  relationships(): IterableIterator<[string, StandardRelationship]> {
    return this._relationships.entries()
  }

  annotations(): IterableIterator<[string, Pattern<Subject>]> {
    return this._annotations.entries()
  }

  walks(): IterableIterator<[string, Pattern<Subject>]> {
    return this._walks.entries()
  }

  other(): ReadonlyArray<Pattern<Subject>> {
    return [...this._other.values()]
  }

  node(id: string): Option.Option<Pattern<Subject>> {
    return liftOption(this._nodes.get(id))
  }

  relationship(id: string): Option.Option<StandardRelationship> {
    return liftOption(this._relationships.get(id))
  }

  annotation(id: string): Option.Option<Pattern<Subject>> {
    return liftOption(this._annotations.get(id))
  }

  walk(id: string): Option.Option<Pattern<Subject>> {
    return liftOption(this._walks.get(id))
  }

  source(id: string): Option.Option<Pattern<Subject>> {
    return pipe(
      this.relationship(id),
      Option.flatMap((relationship) => this.node(relationship.source))
    )
  }

  target(id: string): Option.Option<Pattern<Subject>> {
    return pipe(
      this.relationship(id),
      Option.flatMap((relationship) => this.node(relationship.target))
    )
  }

  neighbors(nodeId: string): ReadonlyArray<Pattern<Subject>> {
    const neighbors: Array<Pattern<Subject>> = []
    for (const relationship of this._relationships.values()) {
      if (relationship.source === nodeId) {
        const target = this._nodes.get(relationship.target)
        if (target) neighbors.push(target)
      } else if (relationship.target === nodeId) {
        const source = this._nodes.get(relationship.source)
        if (source) neighbors.push(source)
      }
    }
    return neighbors
  }

  degree(nodeId: string): number {
    return this.neighbors(nodeId).length
  }

  private ingest(pattern: Pattern<Subject>): void {
    switch (classifyPattern(pattern)) {
      case "node":
        this._nodes.set(pattern.value.identity, pattern)
        break
      case "relationship": {
        const [sourcePattern, targetPattern] = pattern.elements
        if (sourcePattern && targetPattern) {
          this.ingest(sourcePattern)
          this.ingest(targetPattern)
          this._relationships.set(pattern.value.identity, {
            pattern,
            source: sourcePattern.value.identity,
            target: targetPattern.value.identity,
          })
        }
        break
      }
      case "walk":
        for (const element of pattern.elements) {
          this.ingest(element)
        }
        this._walks.set(pattern.value.identity, pattern)
        break
      case "annotation": {
        const [inner] = pattern.elements
        if (inner) {
          this.ingest(inner)
        }
        this._annotations.set(pattern.value.identity, pattern)
        break
      }
      case "other":
        this._other.set(pattern.value.identity, pattern)
        break
    }
  }
}

type PatternClass = "node" | "relationship" | "annotation" | "walk" | "other"

function classifyPattern(pattern: Pattern<Subject>): PatternClass {
  if (pattern.elements.length === 0) {
    return "node"
  }
  if (pattern.elements.length === 1) {
    return "annotation"
  }
  if (pattern.elements.length === 2 && pattern.elements.every(isNodeLike)) {
    return "relationship"
  }
  if (pattern.elements.length >= 1 && pattern.elements.every(isRelationshipLike) && isValidWalk(pattern.elements)) {
    return "walk"
  }
  return "other"
}

function isNodeLike(pattern: Pattern<Subject>): boolean {
  return pattern.elements.length === 0
}

function isRelationshipLike(pattern: Pattern<Subject>): boolean {
  return pattern.elements.length === 2 && pattern.elements.every(isNodeLike)
}

function isValidWalk(patterns: ReadonlyArray<Pattern<Subject>>): boolean {
  if (patterns.length === 0) {
    return false
  }

  let active: Array<Pattern<Subject>> = []
  for (const pattern of patterns) {
    const [left, right] = pattern.elements
    if (!left || !right) {
      return false
    }

    if (active.length === 0) {
      active = [left, right]
      continue
    }

    const next: Array<Pattern<Subject>> = []
    if (active.some((candidate) => candidate.value.identity === left.value.identity)) {
      next.push(right)
    }
    if (active.some((candidate) => candidate.value.identity === right.value.identity)) {
      next.push(left)
    }
    active = next
  }

  return active.length > 0
}

function liftOption<A>(value: A | undefined): Option.Option<A> {
  return value === undefined ? Option.none() : Option.some(value)
}

File: python/packages/relateby/relateby/pattern/_standard_graph.py
"""Native Python StandardGraph.

Classifies Pattern[Subject] values into nodes, relationships, annotations,
walks, and other patterns using the same shape rules as the gram-hs reference.
"""
from __future__ import annotations

from typing import Iterator, Optional

from relateby.gram import parse_gram

from ._pattern import Pattern
from ._subject import Subject


class StandardGraph:
    """An immutable-style graph built by classifying ``Pattern[Subject]`` values.

    Patterns are classified into nodes, relationships, annotations,
    walks, and other according to the same shape rules used by the
    gram-hs reference implementation.  Once ingested, elements are
    accessible via identity-keyed lookups or typed iterators.

    Use the class methods ``from_patterns`` or ``from_gram`` to
    construct a graph rather than calling ``__init__`` directly.
    """

    def __init__(self) -> None:
        self._nodes: dict[str, Pattern[Subject]] = {}
        self._relationships: dict[str, dict[str, object]] = {}
        self._annotations: dict[str, Pattern[Subject]] = {}
        self._walks: dict[str, Pattern[Subject]] = {}
        self._other: dict[str, Pattern[Subject]] = {}

    @classmethod
    def from_patterns(cls, patterns: list[Pattern[Subject]]) -> "StandardGraph":
        """Build a StandardGraph from a list of ``Pattern[Subject]`` values.

        Each pattern is classified (node, relationship, annotation,
        walk, or other) and its components are recursively ingested.

        Args:
            patterns: The patterns to ingest.

        Returns:
            A new StandardGraph containing all classified elements.
        """
        graph = cls()
        for pattern in patterns:
            graph._ingest(pattern)
        return graph

    @classmethod
    def from_gram(cls, input: str) -> "StandardGraph":
        """Parse Gram notation and build a StandardGraph.

        Args:
            input: A string of Gram notation.

        Returns:
            A new StandardGraph built from the parsed patterns.
        """
        return cls.from_patterns(parse_gram(input))

    @property
    def node_count(self) -> int:
        """Number of node patterns currently stored in the graph."""
        return len(self._nodes)

    @property
    def relationship_count(self) -> int:
        """Number of relationship patterns currently stored in the graph."""
        return len(self._relationships)

    @property
    def annotation_count(self) -> int:
        """Number of annotation patterns currently stored in the graph."""
        return len(self._annotations)

    @property
    def walk_count(self) -> int:
        """Number of walk patterns currently stored in the graph."""
        return len(self._walks)

    @property
    def is_empty(self) -> bool:
        """True when the graph contains no classified or other elements."""
        return (
            self.node_count == 0
            and self.relationship_count == 0
            and self.annotation_count == 0
            and self.walk_count == 0
            and len(self._other) == 0
        )

    @property
    def has_conflicts(self) -> bool:
        """True when the graph contains conflicting elements (always False in this implementation)."""
        return False

    def nodes(self) -> Iterator[tuple[str, Pattern[Subject]]]:
        """Iterate over ``(identity, pattern)`` pairs for all node elements."""
        return iter(self._nodes.items())

    def relationships(self) -> Iterator[tuple[str, dict[str, object]]]:
        """Iterate over ``(identity, rel_data)`` pairs for all relationship elements.

        Each ``rel_data`` dict has keys ``"pattern"``, ``"source"``, and
        ``"target"``.
        """
        return iter(self._relationships.items())

    def annotations(self) -> Iterator[tuple[str, Pattern[Subject]]]:
        """Iterate over ``(identity, pattern)`` pairs for all annotation elements."""
        return iter(self._annotations.items())

    def walks(self) -> Iterator[tuple[str, Pattern[Subject]]]:
        """Iterate over ``(identity, pattern)`` pairs for all walk elements."""
        return iter(self._walks.items())

    def other(self) -> list[Pattern[Subject]]:
        """Return a list of patterns that did not match any standard classification."""
        return list(self._other.values())

    def node(self, id: str) -> Optional[Pattern[Subject]]:
        """Look up a node pattern by identity.

        Args:
            id: The identity of the node to retrieve.

        Returns:
            The matching node Pattern, or None if not found.
        """
        return self._nodes.get(id)

    def relationship(self, id: str) -> Optional[dict[str, object]]:
        """Look up relationship data by identity.

        Args:
            id: The identity of the relationship to retrieve.

        Returns:
            A dict with keys ``"pattern"``, ``"source"``, and
            ``"target"``, or None if not found.
        """
        return self._relationships.get(id)

    def annotation(self, id: str) -> Optional[Pattern[Subject]]:
        """Look up an annotation pattern by identity.

        Args:
            id: The identity of the annotation to retrieve.

        Returns:
            The matching annotation Pattern, or None if not found.
        """
        return self._annotations.get(id)

    def walk(self, id: str) -> Optional[Pattern[Subject]]:
        """Look up a walk pattern by identity.

        Args:
            id: The identity of the walk to retrieve.

        Returns:
            The matching walk Pattern, or None if not found.
        """
        return self._walks.get(id)

    def source(self, rel_id: str) -> Optional[Pattern[Subject]]:
        """Return the source node pattern of a relationship.

        Args:
            rel_id: The identity of the relationship.

        Returns:
            The source node Pattern, or None if the relationship or its
            source node is not found.
        """
        relationship = self.relationship(rel_id)
        if relationship is None:
            return None
        return self.node(str(relationship["source"]))

    def target(self, rel_id: str) -> Optional[Pattern[Subject]]:
        """Return the target node pattern of a relationship.

        Args:
            rel_id: The identity of the relationship.

        Returns:
            The target node Pattern, or None if the relationship or its
            target node is not found.
        """
        relationship = self.relationship(rel_id)
        if relationship is None:
            return None
        return self.node(str(relationship["target"]))

    def neighbors(self, node_id: str) -> list[Pattern[Subject]]:
        """Return all node patterns adjacent to the given node (undirected).

        A node is considered adjacent if it appears as the source or
        target of any relationship whose other endpoint is ``node_id``.

        Args:
            node_id: The identity of the node whose neighbors are wanted.

        Returns:
            A list of adjacent node Patterns (may be empty).
        """
        neighbors: list[Pattern[Subject]] = []
        for relationship in self._relationships.values():
            if relationship["source"] == node_id:
                neighbor = self.node(str(relationship["target"]))
                if neighbor is not None:
                    neighbors.append(neighbor)
            elif relationship["target"] == node_id:
                neighbor = self.node(str(relationship["source"]))
                if neighbor is not None:
                    neighbors.append(neighbor)
        return neighbors

    def degree(self, node_id: str) -> int:
        """Return the number of adjacent nodes (undirected degree).

        Args:
            node_id: The identity of the node.

        Returns:
            The count of distinct neighbors (incident relationship
            endpoints on the other side).
        """
        return len(self.neighbors(node_id))

    def _ingest(self, pattern: Pattern[Subject]) -> None:
        classification = classify_pattern(pattern)
        if classification == "node":
            self._nodes[pattern.value.identity] = pattern
        elif classification == "relationship":
            source_pattern = pattern.elements[0]
            target_pattern = pattern.elements[1]
            self._ingest(source_pattern)
            self._ingest(target_pattern)
            self._relationships[pattern.value.identity] = {
                "pattern": pattern,
                "source": source_pattern.value.identity,
                "target": target_pattern.value.identity,
            }
        elif classification == "walk":
            for element in pattern.elements:
                self._ingest(element)
            self._walks[pattern.value.identity] = pattern
        elif classification == "annotation":
            inner = pattern.elements[0]
            self._ingest(inner)
            self._annotations[pattern.value.identity] = pattern
        else:
            self._other[pattern.value.identity] = pattern


def classify_pattern(pattern: Pattern[Subject]) -> str:
    if len(pattern.elements) == 0:
        return "node"
    if len(pattern.elements) == 1:
        return "annotation"
    if len(pattern.elements) == 2 and all(is_node_like(element) for element in pattern.elements):
        return "relationship"
    if (
        len(pattern.elements) >= 1
        and all(is_relationship_like(element) for element in pattern.elements)
        and is_valid_walk(pattern.elements)
    ):
        return "walk"
    return "other"


def is_node_like(pattern: Pattern[Subject]) -> bool:
    return len(pattern.elements) == 0


def is_relationship_like(pattern: Pattern[Subject]) -> bool:
    return len(pattern.elements) == 2 and all(is_node_like(element) for element in pattern.elements)


def is_valid_walk(patterns: list[Pattern[Subject]]) -> bool:
    if not patterns:
        return False

    active: list[Pattern[Subject]] = []
    for pattern in patterns:
        if len(pattern.elements) != 2:
            return False

        left, right = pattern.elements
        if not active:
            active = [left, right]
            continue

        next_active: list[Pattern[Subject]] = []
        if any(candidate.value.identity == left.value.identity for candidate in active):
            next_active.append(right)
        if any(candidate.value.identity == right.value.identity for candidate in active):
            next_active.append(left)
        active = next_active

    return len(active) > 0
